画像判定処理機能を追加

kobayashi-m42 opened the pull request:
# issueURL

https://github.com/nekochans/lgtm-cat-processor/issues/56

# この PR で対応する範囲 / この PR で対応しない範囲

## 対応する範囲

- JudgeImageUsecaseの画像判定処理の実装
- Cognito認証を利用したAPI呼び出し用トークン取得処理
- 画像判定APIを呼び出して画像の適切性を判定する処理
- 判定結果が適切な場合にS3へ画像をコピーする処理

## 対応しない範囲

- StepFunctionsのステートマシンの修正（lgtm-cat-terraform で対応）

# 変更点概要

lgtm-cat-apiの画像判定APIを利用して、アップロードされた画像が適切かどうかを判定する機能を追加した。

判定結果が「適切」の場合はS3コピーでアップロード先バケットに転送し、「不適切」の場合は例外をスローして処理を異常終了させる。

# 補足情報

## Lambdaに追加で設定が必要な環境変数

| 環境変数 | 用途 |
|---------|------|
| `JUDGE_IMAGE_UPLOAD_BUCKET` | アップロード先S3バケット名 |
| `JUDGE_IMAGE_API_URL` | 画像判定APIのベースURL |
| `COGNITO_TOKEN_ENDPOINT` | Cognitoトークンエンドポイント |
| `COGNITO_CLIENT_ID` | CognitoクライアントID |
| `COGNITO_CLIENT_SECRET` | Cognitoクライアントシークレット |

## 関連作業

環境変数の追加とStepFunctionsのステートマシンの修正は https://github.com/nekochans/lgtm-cat-terraform/issues/161 で対応する。

kobayashi-m42 committed:
画像判定処理機能を追加

Cognito認証を経由して外部APIで猫画像の妥当性を検証し、適切な場合はS3バケット間で画像をコピーする処理を実装した。

kobayashi-m42 committed:
コーディング規約をドキュメントに追加

開発中にインポートパスやTypedDictの型ヒントに関する一貫性のない記述が見られたため、チーム全体で統一したコーディングスタイルを維持するためのルールを明文化した。

kobayashi-m42 committed:
認証・画像判定リポジトリのDI実装をクリーンアーキテクチャに統一

kobayashi-m42 commented:
@coderabbitai review

keitakn commented on src/infrastructure/cognito_auth_repository.py:
```diff
@@ -0,0 +1,82 @@
+        self._cached_token: str | None = None
+        self._token_expires_at: float = 0.0
+
+    def get_access_token(self) -> str:
```
![badge](https://img.shields.io/badge/review-imo-orange.svg) 

確かにCacheからトークンを返しているパターンもあるけど、トークンの新規発行もあるから `request_access_token` みたいな名前が良いと思う🐱

もうちょっと正確に言うとトークン発行のクレデンシャルフローは複数あるから一番正確に言うなら `request_access_token_with_client_credentials` だけど `request_access_token` で問題はないと思う！

ちなみにOAuth 2.0 の RFC 6749だと `the authorization server issues an access token …` のように `アクセストークンを発行` みたいな単語がよく出てくるので `issue_access_token`  もアリかもと思ったけど、これは自分達がOAuthとかOpenIDConnectのプロバイダ側でアクセストークンをクライアントに発行する立場なら自然な命名だけど、今のケースだとプロバイダ側であるCognitoに対してアクセストークンを要求リクエストを送信するだからやはり `request_access_token` が適切だと思った！

kobayashi-m42 replied:
@keitakn 
コメントありがとう！確かに、`request_access_token`の方が適切だと思ったから以下で修正したよ！

[認証リポジトリのメソッド名をget_access_tokenからrequest_access_tokenに変更](https://github.com/nekochans/lgtm-cat-processor/pull/58/commits/d89d77406d7e9fd56621e5dd97122fc31aff0ea9)

keitakn commented on pyproject.toml:
```diff
@@ -11,6 +11,7 @@ dependencies = [
     "boto3-stubs[s3,rekognition,s3vectors,bedrock-runtime]>=1.40.69",
     "sqlalchemy>=2.0.44",
     "mysql-connector-python>=9.5.0",
+    "requests>=2.32.0",
```
![badge](https://img.shields.io/badge/review-ask-blue.svg) 

一応確認だけど `aiohttp` じゃなくて `requests` を使っている理由は全体的に非同期関数ではなく同期関数を使っているからで合っている🐈️？

ちなみにAWS LambdaだからそもそもAWS Lambda関数自体が非同期処理の移譲みたいな物だし、無理に非同期にする必要はないと思ってる！

kobayashi-m42 replied:
@keitakn 

> 一応確認だけど aiohttp じゃなくて requests を使っている理由は全体的に非同期関数ではなく同期関数を使っているからで合っている🐈️？

この認識だったよ！ `aiohttp`で非同期処理にする必要はないと思って`requests`を採用しているよ🐱

keitakn replied:
@kobayashi-m42 そうだよね、であれば問題ないと思う👍️

keitakn reviewed (commented):
@kobayashi-m42 

ありがとう🐈️

2個だけコメント（1個は一応確認のコメント）しておいた👍️

kobayashi-m42 committed:
認証リポジトリのメソッド名をget_access_tokenからrequest_access_tokenに変更

OAuth 2.0のクライアントクレデンシャルフローにおいて、プロバイダ（Cognito）に対してアクセストークンを「要求」する 立場であることを明確にする命名に修正

keitakn reviewed (approved):
@kobayashi-m42 

ありがとう🐱問題なさそう👍️

[![LGTMeow](https://lgtm-images.lgtmeow.com/2024/09/22/21/8f35d9c2-2598-49e5-9555-fde45ddf8eab.webp)](https://lgtmeow.com)
